*: Add 'WithOverride' when merging configs

adityasaky opened the pull request:
Possible fix for #508

pjbgf commented:
@adityasaky please add a unit test to ensure in the future we don't regress this behaviour. Also, re-word the commit in line with the contributing guidelines - as per PR title.

pjbgf commented:
@adityasaky apologies for the delay getting back to you. I would simply refactor that code so that the actual merge happens on a func that receives two different configs. Then the test only focus on the new func.

adityasaky commented:
@pjbgf, I moved the merge into a helper. I also noticed that because we're overriding values, we must reverse the direction of overrides. So `local` is merged into `global`, overriding any collisions with the `local` value, which is then merged into `system`. I also added some tests based on how the rest are implemented, though I haven't used gopkg.in/check.v1 before, so I've probably not done it right.

Locally, I'm having issues with the last test case. I'm unable to tell if I've got something wrong with the test case or if I'm invoking the wrong checker, as the behavior seems right to me based off https://go.dev/play/p/XgyYmDsgKL7. Any thoughts?

adityasaky commented:
After some more tweaking, I realized the issue is with how slices are overwritten / merged. Life would get simpler if plumbing/format/config.Option were just a map[string]string instead of a struct, as the issue is with merging config.Options. mergo provides some options for merging slices but none of them really address this as it's hard to know the semantics of merging arbitrary slices...
